src/gossip/service.zig

```diff
@@ -258,12 +260,40 @@ pub const GossipService = struct {
         deinitMux(&self.failed_pull_hashes_mux);
     }
 
-    /// these threads should run forever - so if they join - somethings wrong
-    /// and we should shutdown
-    fn joinAndExit(self: *Self, handle: *std.Thread) void {
-        handle.join();
-        self.exit.store(true, .unordered);
-    }
+    pub const RunHandles = struct {
+        exit: *AtomicBool,
+        receiver_thread: std.Thread,
+        packet_verifier_thread: std.Thread,
+        message_processor_thread: std.Thread,
+        message_builder_thread: ?std.Thread,
+        responder_thread: std.Thread,
+        dumper_thread: ?std.Thread,
+
+        /// If any of the threads join, all other threads will be signalled to join.
+        pub fn joinAndExit(handles: RunHandles) void {
+            inline for (@typeInfo(RunHandles).Struct.fields, 0..) |field, i| cont: {
+                comptime if (@field(std.meta.FieldEnum(RunHandles), field.name) == .exit) {
+                    std.debug.assert(field.type == *AtomicBool);
+                    continue;
+                };
+                const maybe_thread: ?std.Thread = @field(handles, field.name);
+                const thread = maybe_thread orelse break :cont;
+                thread.join(); // if we end up joining, something's gone wrong, so signal exit
+                if (i == 0) handles.exit.store(true, .unordered);
+            }
+        }
+    };
+
+    pub const RunThreadsParams = struct {
```

Comment:
At the relevant callsites, I kept getting confused on as to the order of `spy_node` and `dump` as parameters, so I decided to assign them names. Another solution would be to use an enum instead of a boolean, I just opted for the path of least resistance.

Reply:
gotcha - yeah that makes sense - thinking if we should adopt this going forward 🤔  - i like the explicit names for reading too

Reply:
Yeah, struct param for named parameters, and descriptive enums are both conventions I would like to see used in the codebase, unnamed booleans and integers can be pretty hard to follow otherwise.